From bd77070f6c17a0dba479ed5a45b940d8605a91f3 Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Fri, 17 Jul 2026 11:08:42 +0200
Subject: [PATCH] evaluate: do not fetch next expression on runaway number of
 concatenation components

JIRA: https://issues.redhat.com/browse/RHEL-190549
Upstream Status: nftables commit 955bb6d31c90453e43043346c917646ddc4e5c4e

commit 955bb6d31c90453e43043346c917646ddc4e5c4e
Author: Pablo Neira Ayuso <pablo@netfilter.org>
Date:   Wed Jan 10 18:20:47 2024 +0100

    evaluate: do not fetch next expression on runaway number of concatenation components

    If this is the last expression, then the runaway flag is set on and
    evaluation bails in the next iteration, do not fetch next list element
    which refers to the list head.

    I found this by code inspection, I could not trigger any crash with this
    one.

    Fixes: ae1d54d1343f ("evaluate: do not crash on runaway number of concatenation components")
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 src/evaluate.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/evaluate.c b/src/evaluate.c
index 8f98b48..079f340 100644
--- a/src/evaluate.c
+++ b/src/evaluate.c
@@ -1591,8 +1591,8 @@ static int expr_evaluate_concat(struct eval_ctx *ctx, struct expr **expr)
 		if (key && expressions) {
 			if (list_is_last(&key->list, expressions))
 				runaway = true;
-
-			key = list_next_entry(key, list);
+			else
+				key = list_next_entry(key, list);
 		}
 
 		ctx->inner_desc = NULL;
